packages/js/src/utils/create-async-iterable/create-async-iteratable.ts

```diff
@@ -0,0 +1,67 @@
+export interface AsyncPushCallbacks<T> {
+  next: (value: T) => void;
+  done: () => void;
+  error: (err: unknown) => void;
+}
+
+export function createAsyncIterable<T = unknown>(
+  listener: (ls: AsyncPushCallbacks<T>) => void
+): AsyncIterable<T> {
+  let done = false;
+  let error: unknown | null = null;
+
+  const pushQueue: T[] = [];
+  const pullQueue: Array<
+    [
+      (x: { value: T | undefined; done: boolean }) => void,
+      (err: unknown) => void
+    ]
+  > = [];
+
+  return {
+    [Symbol.asyncIterator]() {
+      listener({
+        next: (value) => {
+          if (done || error) return;
+          if (pullQueue.length > 0) {
+            pullQueue.shift()?.[0]({ value, done: false });
+          } else {
+            pushQueue.push(value);
+          }
+        },
+        error: (err) => {
+          if (done || error) return;
+          if (pullQueue.length > 0) {
+            pullQueue.shift()?.[1](err);
+          } else {
+            error = err;
```

Comment:
I think this has to happen always. Otherwise, the error won't be saved and the iterable will continue being in a non-error state

Reply:
👍